lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -797,7 +797,159 @@ func (d *zfs) CreateVolumeFromMigration(vol Volume, conn io.ReadWriteCloser, vol
 
 // RefreshVolume updates an existing volume to match the state of another.
 func (d *zfs) RefreshVolume(vol Volume, srcVol Volume, srcSnapshots []Volume, op *operations.Operation) error {
-	return genericVFSCopyVolume(d, nil, vol, srcVol, srcSnapshots, true, op)
+	// Get target snapshots
+	targetSnapshots, err := vol.Snapshots(op)
+	if err != nil {
+		return fmt.Errorf("Failed to get target snapshots: %w", err)
+	}
+
+	srcSnapshotsAll, err := srcVol.Snapshots(op)
+	if err != nil {
+		return err
+	}
+
+	// If there are no target or source snapshots, perform a simple copy using zfs.
+	// We cannot use generic vfs volume copy here, as zfs will complain if a generic
+	// copy/refresh is followed by an optimized refresh.
+	if len(targetSnapshots) == 0 || len(srcSnapshotsAll) == 0 {
```

Comment:
Maybe the comment isn't good enough. Consider the following code:
```
lxc init images:alpine/edge c1
lxc copy c1 c2
lxc snapshot c1
lxc copy c1 c2 --refresh
lxc copy c1 c2 --refresh
```
The last command will fail if we used `genericVFSCopyVolume`. The error is
```
cannot receive incremental stream: most recent snapshot of zfs1/containers/c2 does not
match incremental source
```
That's because c2 needs to be a "proper" copy of c1 to make the refresh work correctly.